File: week-2/assignment-2/part-2/section-3/memory.cpp
#include <iostream>
#include <vector>

class MemoryDemo {
private:
    std::vector<int>* data;
    static int instanceCount;
    int id;

public:
    MemoryDemo() {
        data = new std::vector<int>();
        id = ++instanceCount;
        std::cout << "Created MemoryDemo instance " << id << std::endl;
    }

    void addData(int value) {
        data->push_back(value);
    }

    void printData() const {
        std::cout << "Data in instance " << id << ": ";
        for (int val : *data) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }

    ~MemoryDemo() {
        delete data;
        std::cout << "Deleted MemoryDemo instance " << id << " and freed memory." << std::endl;
    }
};

int MemoryDemo::instanceCount = 0;

int main() {
    MemoryDemo* demo1 = new MemoryDemo();
    demo1->addData(5);
    demo1->addData(10);
    demo1->printData();

    {
        MemoryDemo* demo2 = new MemoryDemo();
        demo2->addData(20);
        demo2->printData();
        delete demo2;  // Manual deallocation
    }

    std::cout << "Back to main." << std::endl;

    delete demo1;  // Manual deallocation

    return 0;
}

File: week-2/assignment-2/part-2/section-3/memory.rs
use std::collections::VecDeque;

struct MemoryDemo {
    data: VecDeque<i32>,
}

impl MemoryDemo {
    fn new() -> MemoryDemo {
        println!("Creating a new MemoryDemo instance.");
        MemoryDemo {
            data: VecDeque::new(),
        }
    }
    fn add_data(&mut self, value: i32) {
        self.data.push_back(value);
    }
    fn print_data(&self) {
        println!("Current data: {:?}", self.data);
    }
}

impl Drop for MemoryDemo {
    fn drop(&mut self) {
        println!("MemoryDemo instance is being dropped and memory is freed.");
    }
}

fn main() {
    let mut demo = MemoryDemo::new();
    demo.add_data(5);
    demo.add_data(10);
    demo.print_data();

    {
        let mut demo2 = MemoryDemo::new();
        demo2.add_data(20);
        demo2.print_data();
        // demo2 goes out of scope here, memory is automatically freed
    }

    println!("Back to main.");
}
